src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/StorePal.AndroidKeyStore.cs

```diff
@@ -0,0 +1,149 @@
+            {
+                _keyStoreHandle = keyStoreHandle;
+                _readOnly = (openFlags & (OpenFlags.ReadWrite | OpenFlags.MaxAllowed)) == 0;
+            }
+
+            public SafeHandle SafeHandle => _keyStoreHandle;
+
+            public void Dispose()
+            {
+                _keyStoreHandle.Dispose();
+            }
+
+            public void Add(ICertificatePal cert)
+            {
+                if (_readOnly)
+                    throw new CryptographicException(SR.Cryptography_X509_StoreReadOnly);
+
+                AndroidCertificatePal certPal = (AndroidCertificatePal)cert;
+
+                bool success;
+                if (certPal.HasPrivateKey)
+                {
+                    Interop.AndroidCrypto.PAL_KeyAlgorithm algorithm = certPal.PrivateKeyHandle switch
+                    {
+                        // The AndroidKeyStore doesn't support adding DSA private key entries in newer versions (API 23+)
+                        // Our minimum supported version (API 21) does support it, but for simplicity, we simply block adding
+                        // certificates with DSA private keys on all versions instead of trying to support it on two versions.
+                        SafeDsaHandle _ => throw new PlatformNotSupportedException(SR.Cryptography_X509_StoreDSAPrivateKeyNotSupported),
+                        SafeEcKeyHandle _ => Interop.AndroidCrypto.PAL_KeyAlgorithm.EC,
+                        SafeRsaHandle _ => Interop.AndroidCrypto.PAL_KeyAlgorithm.RSA,
+                        _ => throw new NotSupportedException(SR.NotSupported_KeyAlgorithm)
+                    };
+
+                    success = Interop.AndroidCrypto.X509StoreAddCertificateWithPrivateKey(_keyStoreHandle, certPal.SafeHandle, certPal.PrivateKeyHandle, algorithm);
+                }
+                else
+                {
+                    success = Interop.AndroidCrypto.X509StoreAddCertificate(_keyStoreHandle, certPal.SafeHandle);
+                }
+
+                if (!success)
+                    throw new CryptographicException(SR.Cryptography_X509_StoreAddFailure);
+            }
+
+            public void Remove(ICertificatePal cert)
+            {
+                AndroidCertificatePal certPal = (AndroidCertificatePal)cert;
+                if (_readOnly)
+                {
+                    bool containsCert = Interop.AndroidCrypto.X509StoreContainsCertificate(_keyStoreHandle, certPal.SafeHandle);
+                    if (containsCert)
+                        throw new CryptographicException(SR.Cryptography_X509_StoreReadOnly);
+
+                    // Removing a non-existent certificate is not an error
+                    return;
+                }
+
+                bool success = Interop.AndroidCrypto.X509StoreRemoveCertificate(_keyStoreHandle, certPal.SafeHandle);
+                if (!success)
+                    throw new CryptographicException(SR.Cryptography_X509_StoreRemoveFailure);
```

Comment:
I feel like Windows has cases where remove can look like it should have worked, but doesn't remove something, and no exception is thrown (which is why we don't already have a message).

That said, I'm fine with having an exception here, we can always remove it later.